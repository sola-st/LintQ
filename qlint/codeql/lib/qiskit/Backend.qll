import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.Register
import qiskit.Gate
import qiskit.Circuit

/** Simulator backend. */
class Backend extends DataFlow::CallCfgNode {
  // case: from qiskit import Aer; Aer.get_backend('qasm_simulator')
  Backend() {
    this = API::moduleImport("qiskit").getMember("Aer").getMember("get_backend").getACall() or
    this = API::moduleImport("qiskit").getMember("BasicAer").getMember("get_backend").getACall()
  }

  /** The name of the backend (according to the initialization string). */
  string getKind() {
    // case: from qiskit import Aer; Aer.get_backend('qasm_simulator')
    // > 'qasm_simulator'
    exists(StrConst str, DataFlow::LocalSourceNode strSource |
      strSource = this.getArg(0) and
      strSource.asExpr() = str
    |
      result = str.getText()
    )
  }

  /** If the simulator is statevector simulator. */
  predicate isStatevectorSimulator() {
    // case: from qiskit import Aer; Aer.get_backend('statevector_simulator')
    this.getKind().matches("%statevector%")
  }

  /** If the simulator is Unitary simulator. */
  predicate isUnitarySimulator() {
    // case: from qiskit import Aer; Aer.get_backend('unitary_simulator')
    this.getKind().matches("%unitary%")
  }

  /** Get a circuit run with this backend. */
  QuantumCircuit getACircuitToBeRun() {
    // case: Aer.get_backend('qasm_simulator').run(circuit)
    // > circuit
    exists(QuantumCircuit circuit, DataFlow::CallCfgNode runCall |
      runCall = this.getAnAttributeRead("run").getACall() and
      (
        circuit.flowsTo(runCall.getArg(0)) or
        circuit.flowsTo(runCall.getArgByName("circuits"))
      )
    |
      result = circuit
    )
    or
    // case: execute(qc, Aer.get_backend('qasm_simulator'), shots=1)
    // > qc
    exists(QuantumCircuit circuit, ExecuteCall executeCall |
      (
        circuit.flowsTo(executeCall.getArg(0)) or
        circuit.flowsTo(executeCall.getArgByName("experiments"))
      ) and
      (
        this.flowsTo(executeCall.getArg(1)) or
        this.flowsTo(executeCall.getArgByName("backend"))
      )
    |
      result = circuit
    )
  }

  /** Get a backend run with this backend. */
  BackendRun getABackendRun() {
    exists(BackendRun bkdRun |
      bkdRun.getBackend() = this
    |
      result = bkdRun
    )
  }

}

/** Backend run.
 *
 * This is a call to a backend run method.
 */
abstract class BackendRun extends DataFlow::CallCfgNode {

  abstract Backend getBackend();

}

/** A call to a backend run method. */
class BackendRunViaRunCall extends BackendRun {
  BackendRunViaRunCall() {
    exists(Backend backend, DataFlow::CallCfgNode runCall |
      runCall = backend.getAnAttributeRead("run").getACall()
    |
      this = runCall
    )
  }

  /** The backend that is run. */
  override Backend getBackend() {
    exists(Backend backend, DataFlow::CallCfgNode runCall |
      runCall = backend.getAnAttributeRead("run").getACall()
    |
      result = backend
    )
  }
}

/** A call to a backend execute method. */
class BackendRunViaExecuteCall extends BackendRun {
  BackendRunViaExecuteCall() {
    this instanceof ExecuteCall
  }

  /** The backend that is run. */
  override Backend getBackend() {
    exists(Backend backend |
      (
        backend.flowsTo(this.getArg(1)) or
        backend.flowsTo(this.getArgByName("backend"))
      )
    |
      result = backend
    )
  }
}

/** Result of a simulator run. */
class BackendResult extends DataFlow::CallCfgNode {
  BackendResult() {
    exists(BackendRun bkdRun |
      this = bkdRun.getAnAttributeRead("result").getACall()
      or
      this = bkdRun.getAnAttributeReference("result")
      or
      this = bkdRun.getAnAttributeWrite("result")
    )
  }

  /** The backend run that generated this result. */
  BackendRun getBackendRun() {
    exists(BackendRun bkdRun |
      // connect the run and its result
      this = bkdRun.getAnAttributeRead("result").getACall() or
      this = bkdRun.getAnAttributeReference("result") or
      this = bkdRun.getAnAttributeWrite("result")
    |
      result = bkdRun
    )
  }

}


/** Statevector returned by a result of a simulator run. */
class Statevector extends DataFlow::CallCfgNode {
  Statevector() {
    exists(BackendResult bkdResult |
      this = bkdResult.getAnAttributeRead("get_statevector").getACall()
    )
  }

  /** The backend run that generated this statevector. */
  BackendRun producedByBackendRun() {
    exists(BackendResult bkdResult, BackendRun bkdRun |
      // connect the run and its result
      bkdResult.getBackendRun() = bkdRun and
      // the result has been called to extract the statevector
      this = bkdResult.getAnAttributeRead("get_statevector").getACall()
    |
      result = bkdRun
    )
  }

}