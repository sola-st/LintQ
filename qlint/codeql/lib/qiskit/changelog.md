2023.03.07
- fix missing gates `crx` and `cry` in the `Gate` class
- improved the `isAppliedAfter` in the `Gate` class to remove false positives when the circuit is re-initialized during the path connecting the two gates.

2023.03.07
- added the getVar() function to the register class to get the variable name of the register.
- added the QubitUsed abstraction to keep track of each access to a qubit, considering also the register it belongs to.

2023.03.03
- refactoring model: java style method calls
- refactoring of register to share part of it.
- removed GateQuantumOperation, substituted with a predicate isMeasurement() on a general gate.

2023.02.28
- refactoring of the Gate class. Remove the tracking of the for loop.

2023.02.21
- improve the efficiency of get_a_target_qubit() of the gate class (remove extra QuantumCircuit variable)
- improved the precision of get_a_target_qubit() of the gate class to account also for qubits which are operated on in a loop (e.g. qc.h(i)).


2023.02.20
- added is_subcircuit() method to Circuit class (by checking if it flows to the compose() or append() method of another circuit object)
- improved get_total_num_qubits() and get_total_num_bits() methods of Circuit class (by considering also the registers added add_register() method)
- in Circuit refactored get_num_qubits_with_integers() and get_num_bits_with_integers() methods to return the static size of the circuit (i.e. QuantumCircuit(2,2) or n = 7; m = 2, QuantumCircuit(n, m)).
- improved get_num_bits() and get_num_qubits() methods of Register class (by considering also local flows of integer variables, e.g. x = 5; QuantumRegister(x))