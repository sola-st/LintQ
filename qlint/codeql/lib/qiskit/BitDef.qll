import qiskit.Circuit
import qiskit.Register
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs

/** Definition of a qubit or bit. */
abstract class BitDefinition extends DataFlow::LocalSourceNode {
  int getAnIndex() {
    if exists(int i | i = this.getAnIndexIfAny())
    then exists(int i | i = this.getAnIndexIfAny() | result = i)
    else result = -1
  }

  string getARegisterName() {
    if exists(RegisterV2 reg | reg = this.getARegister())
    then exists(RegisterV2 reg | reg = this.getARegister() | result = reg.getName())
    else result = "anonymous register"
  }

  string getACircuitName() {
    if exists(QuantumCircuit circ | circ = this.getACircuit())
    then exists(QuantumCircuit circ | circ = this.getACircuit() | result = circ.getName())
    else result = "anonymous circuit"
  }

  abstract string getTypeName();

  abstract int getAnIndexIfAny();

  abstract RegisterV2 getARegister();

  abstract QuantumCircuit getACircuit();

  boolean equals(BitDefinition other) {
    if
      this.getARegisterName() = other.getARegisterName() and
      this.getAnIndexIfAny() = other.getAnIndexIfAny() and
      this.getACircuitName() = other.getACircuitName() and
      this.getARegister() = other.getARegister() and
      this.getACircuit() = other.getACircuit() and
      this.getTypeName() = other.getTypeName()
    then result = true
    else result = false
  }
}

abstract class QubitDefinition extends BitDefinition {
  override string getTypeName() { result = "qubit" }
}

abstract class ClbitDefinition extends BitDefinition {
  override string getTypeName() { result = "clbit" }
}

predicate isIntegerParameterOfCall(
  DataFlow::LocalSourceNode integerParameter, DataFlow::CallCfgNode call, int indexParam
) {
  exists(IntegerLiteral intLit |
    integerParameter.flowsTo(call.getArg(indexParam)) and
    integerParameter.asExpr() = intLit
  )
}

/** Qubit generated implicitly by QuantumCircuit(). */
class ImplicitCircuitQubitDefinition extends BitDefinition, QubitDefinition {
  ImplicitCircuitQubitDefinition() {
    exists(QuantumCircuit qc, DataFlow::LocalSourceNode locSource |
      qc instanceof QuantumCircuitConstructor or
      qc instanceof BuiltinParametrizedCircuitsConstructor
    |
      this = locSource and
      isIntegerParameterOfCall(locSource, qc, 0)
    )
  }

  /** Returns an index of the qubit in the register. */
  override int getAnIndexIfAny() {
    // qc = QuantumCircuit(4, 3)
    // >> [0, 1, 2, 3]
    exists(QuantumCircuit qc, DataFlow::LocalSourceNode locSource, int i |
      qc instanceof QuantumCircuitConstructor or
      qc instanceof BuiltinParametrizedCircuitsConstructor
    |
      this = locSource and
      isIntegerParameterOfCall(locSource, qc, 0) and
      i in [0 .. qc.getNumberOfQubits() - 1] and
      result = i
    )
  }

  override RegisterV2 getARegister() {
    exists(QuantumCircuit qc, DataFlow::LocalSourceNode locSource |
      qc instanceof QuantumCircuitConstructor or
      qc instanceof BuiltinParametrizedCircuitsConstructor
    |
      this = locSource and
      isIntegerParameterOfCall(locSource, qc, 0) and
      result = qc
    )
  }

  override QuantumCircuit getACircuit() {
    exists(QuantumCircuit qc, DataFlow::LocalSourceNode locSource |
      qc instanceof QuantumCircuitConstructor or
      qc instanceof BuiltinParametrizedCircuitsConstructor
    |
      this = locSource and
      isIntegerParameterOfCall(locSource, qc, 0) and
      result = qc
    )
  }
}

/** Clbit generated implicitly by QuantumCircuit(). */
class ImplicitCircuitClbitDefinition extends BitDefinition, ClbitDefinition {
  ImplicitCircuitClbitDefinition() {
    exists(QuantumCircuit qc, DataFlow::LocalSourceNode locSource |
      qc instanceof QuantumCircuitConstructor
    |
      this = locSource and
      isIntegerParameterOfCall(locSource, qc, 1)
    )
  }

  /** Returns an index of the clbit in the register. */
  override int getAnIndexIfAny() {
    // qc = QuantumCircuit(4, 3)
    // >> [0, 1, 2]
    exists(QuantumCircuit qc, DataFlow::LocalSourceNode locSource, int i |
      qc instanceof QuantumCircuitConstructor
    |
      this = locSource and
      isIntegerParameterOfCall(locSource, qc, 1) and
      i in [0 .. qc.getNumberOfClassicalBits() - 1] and
      result = i
    )
  }

  override RegisterV2 getARegister() {
    exists(QuantumCircuit qc, DataFlow::LocalSourceNode locSource |
      qc instanceof QuantumCircuitConstructor
    |
      this = locSource and
      isIntegerParameterOfCall(locSource, qc, 1) and
      result = qc
    )
  }

  override QuantumCircuit getACircuit() {
    exists(QuantumCircuit qc, DataFlow::LocalSourceNode locSource |
      qc instanceof QuantumCircuitConstructor
    |
      this = locSource and
      isIntegerParameterOfCall(locSource, qc, 1) and
      result = qc
    )
  }
}

/** Bits generated implicitly by QuantumRegister() and ClassicalRegister(). */
abstract class ImplicitRegisterBitDefinition extends BitDefinition {
  /** Returns an index of the bit in the register. */
  override int getAnIndexIfAny() {
    exists(int i | i in [0 .. this.(RegisterV2).getSize() - 1] | result = i)
  }

  /** Returns the register of the bit. */
  override RegisterV2 getARegister() { result = this }

  /** Returns the circuit in which the bit is added. */
  override QuantumCircuit getACircuit() {
    // get the circuit from the register
    result = this.getARegister().getACircuit()
  }
}

/** Qubit generated implicitly by QuantumRegister(). */
class ImplicitRegisterQubitDefinition extends ImplicitRegisterBitDefinition, QubitDefinition {
  ImplicitRegisterQubitDefinition() {
    exists(QuantumRegisterV2 qreg, int i | i in [0 .. qreg.getSize() - 1] | this = qreg)
  }
}

/** Clbit generated implicity by ClassicalRegister() */
class ImplicitRegisterClbitDefinition extends ImplicitRegisterBitDefinition, ClbitDefinition {
  ImplicitRegisterClbitDefinition() {
    exists(ClassicalRegisterV2 creg, int i | i in [0 .. creg.getSize() - 1] | this = creg)
  }
}

// EXPLICIT INITIALIZATION OF SINGLE BITS
/** Bits generated by Qubit() or Clbit(). */
abstract class ExplicitSingleBitDefinition extends BitDefinition {
  /** Returns the name of the identifier of the bit. */
  string getName() {
    exists(AssignStmt a |
      a.contains(this.(DataFlow::CallCfgNode).getNode().getNode()) and
      result = a.getATarget().(Name).getId()
    )
  }

  /** Return the index paramer of the current bit. */
  override int getAnIndexIfAny() {
    // single_bit = Clbit(register=creg, index=2)
    result =
      this.(API::CallNode)
          .getParameter(1, "index")
          .getAValueReachingSink()
          .asExpr()
          .(IntegerLiteral)
          .getValue()
  }

  /** Returns the register of the current bit. */
  override RegisterV2 getARegister() {
    // single_bit = Clbit(register=creg, index=2)
    exists(RegisterV2 reg |
      reg.asExpr() =
        this.(API::CallNode).getParameter(0, "register").getAValueReachingSink().asExpr()
    |
      result = reg
    )
  }

  /** Returns the circuit in which the bit is added. */
  override QuantumCircuit getACircuit() {
    // get the circuit from the register
    result = this.getARegister().getACircuit()
  }
}

/** Qubit generated by Qubit(). */
class ExplicitSingleQubitDefinition extends ExplicitSingleBitDefinition, QubitDefinition {
  ExplicitSingleQubitDefinition() {
    this = API::moduleImport("qiskit").getMember("Qubit").getACall()
    or
    this = API::moduleImport("qiskit").getMember("circuit").getMember("Qubit").getACall()
  }
}

/** Clbit generated by Clbit(). */
class ExplicitSingleClbitDefinition extends ExplicitSingleBitDefinition, ClbitDefinition {
  ExplicitSingleClbitDefinition() {
    this = API::moduleImport("qiskit").getMember("Clbit").getACall()
    or
    this = API::moduleImport("qiskit").getMember("circuit").getMember("Clbit").getACall()
  }
}
