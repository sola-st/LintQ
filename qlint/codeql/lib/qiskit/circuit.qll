import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.register


class QuantumCircuit extends DataFlow::CallCfgNode {
    QuantumCircuit() {
        this = API::moduleImport("qiskit").getMember("QuantumCircuit").getACall()
    }

    int get_num_qubits() {
        exists(IntegerLiteral num_qubits |
            num_qubits = this.getArg(0).asExpr() and
            result = num_qubits.getValue()
        ) or
        exists( QuantumRegister qntReg|
            // there is a dataflow between the quantum register and the quantum circuit
            exists(int i |
                qntReg.flowsTo(this.getArg(i)))
            and
            // the number of qubits is the number of qubits in the quantum register
            result = qntReg.get_num_qubits()
        )
    }

    int get_num_bits() {
        exists(IntegerLiteral num_bits |
            num_bits = this.getArg(1).asExpr() and
            result = num_bits.getValue()
        ) or
        exists( ClassicalRegister clsReg|
            // there is a dataflow between the classical register and the quantum circuit
            exists(int i |
                clsReg.flowsTo(this.getArg(i)))
            and
            // the number of bits is the number of bits in the classical register
            result = clsReg.get_num_bits()
        )
    }

    int get_total_num_bits() {
        result = sum(ClassicalRegister clsReg, int i |
            clsReg.flowsTo(this.getArg(i)) |
            clsReg.get_num_bits())
    }

    int get_total_num_qubits() {
        result = sum(QuantumRegister qntReg, int i |
            qntReg.flowsTo(this.getArg(i)) |
            qntReg.get_num_qubits())
    }

}