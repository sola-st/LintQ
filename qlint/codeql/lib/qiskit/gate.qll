import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.circuit

class GateName extends string {
    GateName() {
        this = "x" or
        this = "y" or
        this = "z" or
        this = "h" or
        this = "s" or
        this = "sdg" or
        this = "t" or
        this = "tdg" or
        this = "rx" or
        this = "ry" or
        this = "rz" or
        this = "u1" or
        this = "u2" or
        this = "u3" or
        this = "cx" or
        this = "cy" or
        this = "cz" or
        this = "ch" or
        this = "crz" or
        this = "cu1" or
        this = "cu3" or
        this = "swap" or
        this = "ccx" or
        this = "cswap" or
        this = "rxx" or
        this = "ryy" or
        this = "rzz" or
        this = "rzx" or
        this = "rzz" or
        this = "measure" or
        this = "measure_all"
    }
}

// TODO: improve to support
// from qiskit import QuantumCircuit
// qc = QuantumCircuit(2)
// qc.append(CXGate(), [0, 1])

class GenericGate extends DataFlow::CallCfgNode {

    GenericGate() {
        exists(
            QuantumCircuit circ, GateName a_supported_gate_name|
            this = circ.getAnAttributeRead(a_supported_gate_name).getACall()
        )
    }

    string get_gate_name() {
        exists(
            QuantumCircuit circ, GateName a_supported_gate_name |
            this = circ.getAnAttributeRead(a_supported_gate_name).getACall() |
            result = a_supported_gate_name
        )
    }

    QuantumCircuit get_quantum_circuit() {
        exists(
            QuantumCircuit circ |
            this = circ.getAnAttributeRead(this.get_gate_name()).getACall()|
            result = circ
        )
    }


    int get_a_target_qubit() {
        // TODO: improve to support sequences of qubits or integers
        exists(
            QuantumCircuit circ, int target_qubit, int i |
            this = circ.getAnAttributeRead().getACall() and
            target_qubit = this.getArg(i).asExpr().(IntegerLiteral).getValue()|
            // return a list with only the target qubit
            result = target_qubit
        )
        // handle the case where the arguments are in accessed through a quantum
        // reigster variable like this:
        // q = QuantumRegister(2)
        // qc = QuantumCircuit(q)
        // qc.cx(q[0], q[1])
        or
        exists(
            QuantumCircuit circ,
            QuantumRegister qreg,
            DataFlow::Node nd,
            DataFlow::ExprNode targetSubscript,
            Subscript subscript,
            IntegerLiteral bit,
            int i |
                this = circ.getAnAttributeRead().getACall() and
                qreg.flowsTo(nd) and
                nd.asExpr() = targetSubscript.asExpr() and
                targetSubscript.asExpr() = subscript.getObject() and
                subscript = this.getArg(i).asExpr() and
                bit = subscript.getIndex() |
                result = bit.getValue()
        )
    }

    QuantumRegister get_a_target_qubit_in_register() {
        exists(
            QuantumCircuit circ,
            QuantumRegister qntReg |
                this = circ.getAnAttributeRead().getACall() and
                qntReg = circ.get_a_quantum_register() and
                qntReg.flowsTo(this.getArg(0)) |
                result = qntReg
        )
    }

    predicate follows(GenericGate g) {
        exists(
            QuantumCircuit circ |
                this = circ.getAnAttributeRead().getACall() and
                g = circ.getAnAttributeRead().getACall() and
                not this = g and
                this.get_a_target_qubit() = g.get_a_target_qubit() and
                this.get_quantum_circuit() = g.get_quantum_circuit() and
                g.asCfgNode().strictlyReaches(this.asCfgNode())
        )
    }
}


class Gate extends GenericGate {
    Gate() {
        exists(
            QuantumCircuit circ, GateName a_supported_gate_name |
            a_supported_gate_name != "measure"  and a_supported_gate_name != "measure_all"|
            this = circ.getAnAttributeRead(a_supported_gate_name).getACall()
        )
    }
}

class Measure extends GenericGate {
    Measure() {
        exists(
            QuantumCircuit circ, GateName a_supported_gate_name |
            a_supported_gate_name = "measure" |
            this = circ.getAnAttributeRead(a_supported_gate_name).getACall()
        )
    }

    override int get_a_target_qubit() {
        // note that measure gets qubits only in the first position
        // TODO: improve to support sequences of qubits or integers
        // e.g. qc.measure([0, 1], [0, 1])
        // from documentation Parameters
        // qubit (Union[Qubit, QuantumRegister, int, slice, Sequence[Union[Qubit, int]]]) – qubit to measure.
        // cbit (Union[Clbit, ClassicalRegister, int, slice, Sequence[Union[Clbit, int]]]) – classical bit to place the measurement in.

        // TODO: improve to support entire registers
        // e.g. qc.measure(q, c)
        exists(
            QuantumCircuit circ, int target_qubit|
            this = circ.getAnAttributeRead("measure").getACall() and
            target_qubit = this.getArg(0).asExpr().(IntegerLiteral).getValue()|
            // return a list with only the target qubit
            result = target_qubit
        )
        or
        exists(
            QuantumCircuit circ,
            QuantumRegister qreg,
            DataFlow::Node nd,
            DataFlow::ExprNode targetSubscript,
            Subscript subscript,
            IntegerLiteral bit|
                this = circ.getAnAttributeRead("measure").getACall() and
                qreg.flowsTo(nd) and
                nd.asExpr() = targetSubscript.asExpr() and
                targetSubscript.asExpr() = subscript.getObject() and
                subscript = this.getArg(0).asExpr() and
                bit = subscript.getIndex() |
                result = bit.getValue()
        )
    }
}

class MeasureAll extends GenericGate {
    MeasureAll() {
        exists(
            QuantumCircuit circ, GateName a_supported_gate_name |
            a_supported_gate_name = "measure_all" |
            this = circ.getAnAttributeRead(a_supported_gate_name).getACall()
        )
    }
}