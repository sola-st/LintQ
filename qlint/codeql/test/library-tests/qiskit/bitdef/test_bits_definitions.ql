import python
import qiskit.BitDef

from BitDefinition bitDef
where not bitDef.getLocation().getFile().getAbsolutePath().matches("%site-packages/qiskit/%")
select bitDef,
  "generic bit (type: " + bitDef.getTypeName() + ") " + "in reg: '" + bitDef.getARegisterName() +
    "'," + " index: '" + bitDef.getAnIndex() + "', " + " circuit: '" + bitDef.getACircuitName() +
    "'"
