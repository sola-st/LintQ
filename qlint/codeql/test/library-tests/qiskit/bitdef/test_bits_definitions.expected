| concept_bit_def.py:5:8:5:25 | ControlFlowNode for QuantumRegister() | generic bit (type: qubit) in reg: 'qreg', index: '0',  circuit: 'qc' |
| concept_bit_def.py:5:8:5:25 | ControlFlowNode for QuantumRegister() | generic bit (type: qubit) in reg: 'qreg', index: '1',  circuit: 'qc' |
| concept_bit_def.py:5:8:5:25 | ControlFlowNode for QuantumRegister() | generic bit (type: qubit) in reg: 'qreg', index: '2',  circuit: 'qc' |
| concept_bit_def.py:5:8:5:25 | ControlFlowNode for QuantumRegister() | generic bit (type: qubit) in reg: 'qreg', index: '3',  circuit: 'qc' |
| concept_bit_def.py:5:8:5:25 | ControlFlowNode for QuantumRegister() | generic bit (type: qubit) in reg: 'qreg', index: '4',  circuit: 'qc' |
| concept_bit_def.py:5:8:5:25 | ControlFlowNode for QuantumRegister() | generic bit (type: qubit) in reg: 'qreg', index: '5',  circuit: 'qc' |
| concept_bit_def.py:5:8:5:25 | ControlFlowNode for QuantumRegister() | generic bit (type: qubit) in reg: 'qreg', index: '6',  circuit: 'qc' |
| concept_bit_def.py:5:8:5:25 | ControlFlowNode for QuantumRegister() | generic bit (type: qubit) in reg: 'qreg', index: '7',  circuit: 'qc' |
| concept_bit_def.py:6:8:6:27 | ControlFlowNode for ClassicalRegister() | generic bit (type: clbit) in reg: 'creg', index: '0',  circuit: 'qc' |
| concept_bit_def.py:6:8:6:27 | ControlFlowNode for ClassicalRegister() | generic bit (type: clbit) in reg: 'creg', index: '1',  circuit: 'qc' |
| concept_bit_def.py:6:8:6:27 | ControlFlowNode for ClassicalRegister() | generic bit (type: clbit) in reg: 'creg', index: '2',  circuit: 'qc' |
| concept_bit_def.py:6:8:6:27 | ControlFlowNode for ClassicalRegister() | generic bit (type: clbit) in reg: 'creg', index: '3',  circuit: 'qc' |
| concept_bit_def.py:6:8:6:27 | ControlFlowNode for ClassicalRegister() | generic bit (type: clbit) in reg: 'creg', index: '4',  circuit: 'qc' |
| concept_bit_def.py:6:8:6:27 | ControlFlowNode for ClassicalRegister() | generic bit (type: clbit) in reg: 'creg', index: '5',  circuit: 'qc' |
| concept_bit_def.py:6:8:6:27 | ControlFlowNode for ClassicalRegister() | generic bit (type: clbit) in reg: 'creg', index: '6',  circuit: 'qc' |
| concept_bit_def.py:6:8:6:27 | ControlFlowNode for ClassicalRegister() | generic bit (type: clbit) in reg: 'creg', index: '7',  circuit: 'qc' |
| concept_bit_def.py:9:16:9:29 | ControlFlowNode for Qubit() | generic bit (type: qubit) in reg: 'qreg', index: '0',  circuit: 'qc' |
| concept_bit_def.py:10:19:10:35 | ControlFlowNode for Qubit() | generic bit (type: qubit) in reg: 'anonymous register', index: '-1',  circuit: 'anonymous circuit' |
| concept_bit_def.py:12:14:12:27 | ControlFlowNode for Clbit() | generic bit (type: clbit) in reg: 'creg', index: '2',  circuit: 'qc' |
| concept_bit_def.py:13:17:13:23 | ControlFlowNode for Clbit() | generic bit (type: clbit) in reg: 'anonymous register', index: '-1',  circuit: 'anonymous circuit' |
| concept_bit_def_via_circuit.py:7:21:7:21 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '0',  circuit: 'qc' |
| concept_bit_def_via_circuit.py:7:21:7:21 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '1',  circuit: 'qc' |
| concept_bit_def_via_circuit.py:7:24:7:24 | ControlFlowNode for IntegerLiteral | generic bit (type: clbit) in reg: 'anonymous register', index: '0',  circuit: 'qc' |
| concept_bit_def_via_circuit.py:7:24:7:24 | ControlFlowNode for IntegerLiteral | generic bit (type: clbit) in reg: 'anonymous register', index: '1',  circuit: 'qc' |
| concept_bit_def_via_circuit.py:7:24:7:24 | ControlFlowNode for IntegerLiteral | generic bit (type: clbit) in reg: 'anonymous register', index: '2',  circuit: 'qc' |
| concept_bit_def_via_circuit.py:7:24:7:24 | ControlFlowNode for IntegerLiteral | generic bit (type: clbit) in reg: 'anonymous register', index: '3',  circuit: 'qc' |
| concept_bit_def_via_circuit.py:7:24:7:24 | ControlFlowNode for IntegerLiteral | generic bit (type: clbit) in reg: 'anonymous register', index: '4',  circuit: 'qc' |
| concept_bit_def_via_circuit.py:7:24:7:24 | ControlFlowNode for IntegerLiteral | generic bit (type: clbit) in reg: 'anonymous register', index: '5',  circuit: 'qc' |
| concept_bit_def_via_circuit.py:7:24:7:24 | ControlFlowNode for IntegerLiteral | generic bit (type: clbit) in reg: 'anonymous register', index: '6',  circuit: 'qc' |
| concept_bit_def_via_circuit.py:12:36:12:37 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '0',  circuit: 'qc_mostly_quantum' |
| concept_bit_def_via_circuit.py:12:36:12:37 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '1',  circuit: 'qc_mostly_quantum' |
| concept_bit_def_via_circuit.py:12:36:12:37 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '2',  circuit: 'qc_mostly_quantum' |
| concept_bit_def_via_circuit.py:12:36:12:37 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '3',  circuit: 'qc_mostly_quantum' |
| concept_bit_def_via_circuit.py:12:36:12:37 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '4',  circuit: 'qc_mostly_quantum' |
| concept_bit_def_via_circuit.py:12:36:12:37 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '5',  circuit: 'qc_mostly_quantum' |
| concept_bit_def_via_circuit.py:12:36:12:37 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '6',  circuit: 'qc_mostly_quantum' |
| concept_bit_def_via_circuit.py:12:36:12:37 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '7',  circuit: 'qc_mostly_quantum' |
| concept_bit_def_via_circuit.py:12:36:12:37 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '8',  circuit: 'qc_mostly_quantum' |
| concept_bit_def_via_circuit.py:12:36:12:37 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '9',  circuit: 'qc_mostly_quantum' |
| concept_bit_def_via_circuit.py:12:40:12:40 | ControlFlowNode for IntegerLiteral | generic bit (type: clbit) in reg: 'anonymous register', index: '0',  circuit: 'qc_mostly_quantum' |
| concept_bit_def_via_circuit.py:12:40:12:40 | ControlFlowNode for IntegerLiteral | generic bit (type: clbit) in reg: 'anonymous register', index: '1',  circuit: 'qc_mostly_quantum' |
| concept_bit_def_via_circuit.py:12:40:12:40 | ControlFlowNode for IntegerLiteral | generic bit (type: clbit) in reg: 'anonymous register', index: '2',  circuit: 'qc_mostly_quantum' |
| concept_bit_def_via_circuit.py:19:25:19:25 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '0',  circuit: 'qc' |
| concept_bit_def_via_circuit.py:19:25:19:25 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '1',  circuit: 'qc' |
| concept_bit_def_via_circuit.py:30:23:30:23 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '0',  circuit: 'qc_su2' |
| concept_bit_def_via_circuit.py:30:23:30:23 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '1',  circuit: 'qc_su2' |
| concept_bit_def_via_circuit.py:30:23:30:23 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '2',  circuit: 'qc_su2' |
| concept_bit_def_via_circuit.py:33:26:33:26 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '0',  circuit: 'qc_real' |
| concept_bit_def_via_circuit.py:33:26:33:26 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '1',  circuit: 'qc_real' |
| concept_bit_def_via_circuit.py:33:26:33:26 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '2',  circuit: 'qc_real' |
| concept_bit_def_via_circuit.py:33:26:33:26 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '3',  circuit: 'qc_real' |
| concept_bit_def_via_circuit.py:36:18:36:18 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '0',  circuit: 'qc_tl' |
| concept_bit_def_via_circuit.py:36:18:36:18 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '1',  circuit: 'qc_tl' |
| concept_bit_def_via_circuit.py:36:18:36:18 | ControlFlowNode for IntegerLiteral | generic bit (type: qubit) in reg: 'anonymous register', index: '2',  circuit: 'qc_tl' |
